### PR TITLE
Namespace scoped installation

### LINKED ISSUE #1692: Option for installation with non-admin permissions
With current installation, user needs admin privileges on Kubernetes cluster to install Aperture as we use cluster level resources like CRDs, ClusterRole, ClusterRoleBinding and admission webhooks.

- Agent and Controller as Deployment (Non-admin setup)
  - Agent
    - We will have Agent running as Deployment
    - All the configurations which requires Kubernetes ClusterRole like Service discovery, Auto scaler has to be turned off or restricted to a single namespace
     - We will create all the required Kubernetes resources like Deployment, Service, ConfigMap, Secret etc directly from Helm chart i.e there won't be any Operator in between
     - We have to do config validations from Agent as the Webhook which does this with Operator, won't be available in this case
This installation mode would ideally only support SDK based setup due to above limitations
  - Controller
    - All the above points
    - For policy, we will be using ConfigMaps for now and later can think to add CRUD for policy on Controller

### PR DESCRIPTION
**Pull request by hdkshingala**
### Description of change

This PR contains all the changes required for running agent and controller in namespace scoped:
- Option in Agent and Controller helm chart to install only namespace scoped installation
- aperturectl support for installation in namespace scoped
- Disable Policy CR watcher conditionally in Controller
- Policy Post and Delete APIs for managing policies

Fixes #1692 

##### Checklist

- [x] Tested in playground or other setup
- [x] Documentation is changed or added

Default Playground SS:
image

Aperture Running in namespace scoped SS with Envoy:
image

### Summary by OpenAI

**New Feature:**
- Namespace-scoped installation of Aperture Agent and Controller
- New helm chart options, aperturectl support, and policy APIs
- Conditionally disable Policy CR watcher in the Controller
- Policy Post and Delete APIs for managing policies

> 🎉 Namespace-scoped, we now install,
> With helm charts, APIs, and more at our call.
> Policies managed with ease and grace,
> Aperture's power grows, embracing each namespace. 🚀

### CODE REVIEW

File: pkg/policies/controlplane/policy-service.go [hunk 1]
```diff
@@ -29,3 +35,44 @@ func (s *PolicyService) GetPolicies(ctx context.Context, _ *emptypb.Empty) (*pol
...
+		response.Message = "successfully posted policies"
+	}
+	return response, nil
+}
+
+// DeletePolicy deletes a policy from the system.
+func (s *PolicyService) DeletePolicy(ctx context.Context, policy *policylangv1.DeletePolicyRequest) (*emptypb.Empty, error) {
+	s.etcdTracker.RemoveEvent(notifiers.Key(policy.Name))
+	return &emptypb.Empty{}, nil
+}
```
[the remark this answers is not in the document]
Comment (hdkshingala) [reply pkg/policies/controlplane/policy-service.go:78]: WriteEvent and RemoveEvent don't return errors

File: cmd/aperturectl/cmd/installation/install.go [hunk 2]
```diff
@@ -16,7 +16,7 @@ func init() {
 	InstallCmd.PersistentFlags().StringVar(&kubeConfig, "kube-config", "", "Path to the Kubernetes cluster config. Defaults to '~/.kube/config'")
 	InstallCmd.PersistentFlags().StringVar(&valuesFile, "values-file", "", "Values YAML file containing parameters to customize the installation")
 	InstallCmd.PersistentFlags().StringVar(&version, "version", apertureLatestVersion, "Version of the Aperture")
-	InstallCmd.PersistentFlags().StringVar(&namespace, "namespace", "", "Namespace in which the component will be installed. Defaults to component name")
+	InstallCmd.PersistentFlags().StringVar(&namespace, "namespace", "", "Namespace in which the component will be installed. Defaults to 'default' namespace")
 
 	InstallCmd.AddCommand(controllerInstallCmd)
 	InstallCmd.AddCommand(agentInstallCmd)
```
[the remark this answers is not in the document]
Comment (hdkshingala) [reply cmd/aperturectl/cmd/installation/install.go:22]: istioConfigInstallCmd is already present and added

File: cmd/aperturectl/cmd/installation/utils.go [hunk 3]
```diff
@@ -40,6 +45,7 @@ var (
 	namespace      string
 	kubeClient     client.Client
 	timeout        int
+	generateCert   bool
 )
```
[the remark this answers is not in the document]
Comment (hdkshingala) [reply cmd/aperturectl/cmd/installation/utils.go:49]: It is used at line no 414

File: api/aperture/cmd/v1/cmd.proto [hunk 4]
```diff
@@ -21,6 +22,9 @@ service Controller {
   // duplicating a bit preview.v1.FlowPreviewService to keep controller APIs in one place.
   rpc PreviewFlowLabels(PreviewFlowLabelsRequest) returns (PreviewFlowLabelsControllerResponse);
   rpc PreviewHTTPRequests(PreviewHTTPRequestsRequest) returns (PreviewHTTPRequestsControllerResponse);
+
+  rpc PostPolicies(aperture.policy.language.v1.PostPoliciesRequest) returns (aperture.policy.language.v1.PostPoliciesResponse);
+  rpc PatchPolicies(aperture.policy.language.v1.PostPoliciesRequest) returns (aperture.policy.language.v1.PostPoliciesResponse);
 }
```
[the remark this answers is not in the document]
Comment (hdkshingala) [reply api/aperture/cmd/v1/cmd.proto:28]: The both requests are same so we don't have to duplicate them

File: pkg/policies/controlplane/provider.go [hunk 5]
```diff
@@ -18,14 +30,155 @@ const (
 //   in: body
 //   schema:
 //     "$ref": "#/definitions/JobGroupConfig"
+// - name: cr_watcher
+//   in: body
+//   schema:
+//     "$ref": "#/definitions/CRWatcherConfig"
```
[the remark this answers is not in the document]
Comment (hdkshingala) [reply pkg/policies/controlplane/provider.go:37]: This is needed for swagger

File: pkg/policies/controlplane/crwatcher/provide.go [hunk 6]
```diff
@@ -23,26 +36,38 @@ func (constructor Constructor) Annotate() fx.Option {
...
+	trackersName := config.NameTag(constructor.Name)
+	dynamicConfigTrackersName := config.NameTag(constructor.DynamicConfigName)
+	return fx.Options(fx.Invoke(
 		fx.Annotate(
 			constructor.provideWatcher,
-			fx.ResultTags(name, dynamicConfigName),
+			fx.ParamTags(trackersName, dynamicConfigTrackersName),
 		),
 	))
 }
```
[the remark this answers is not in the document]
Comment (hdkshingala) [reply pkg/policies/controlplane/crwatcher/provide.go:47]: This is intentional as it is no longer providing any other resource

File: pkg/cmd/controller.go [hunk 7]
```diff
@@ -347,3 +350,11 @@ agentsLoop:
...
 	return agents, nil
 }
+
+func (h *Handler) PostPolicies(ctx context.Context, req *policylangv1.PostPoliciesRequest) (*policylangv1.PostPoliciesResponse, error) {
+	return h.policyService.PostPolicies(ctx, req)
+}
+
+func (h *Handler) PatchPolicies(ctx context.Context, req *policylangv1.PostPoliciesRequest) (*policylangv1.PostPoliciesResponse, error) {
+	return h.policyService.PatchPolicies(ctx, req)
+}
```
[the remark this answers is not in the document]
Comment (hdkshingala) [reply pkg/cmd/controller.go:360]: This won't be nil as we never return nil while constructing PolicyService so no need of that check

File: operator/controllers/controller/secrets.go [hunk 8]
```diff
@@ -57,7 +57,7 @@ func secretForControllerAPIKey(instance *controllerv1alpha1.Controller, scheme *
 	return secret, nil
 }
 
-// secretForControllerApiKey prepares the Secret object for the ApiKey of Agent.
+// secretForControllerCert prepares the Secret object for the Aperture Controller Certificate.
 func secretForControllerCert(instance *controllerv1alpha1.Controller, scheme *runtime.Scheme, serverCert, serverKey *bytes.Buffer) (*corev1.Secret, error) {
```
[the remark this answers is not in the document]
Comment (hdkshingala) [reply operator/controllers/controller/secrets.go:61]: I already did that change

File: operator/controllers/controller/reconciler.go [hunk 9]
```diff
@@ -330,6 +330,7 @@ func (r *ControllerReconciler) manageResources(ctx context.Context, log logr.Log
 		Enabled:  true,
 	}
 
+	instance.Spec.ConfigSpec.Policies.CRWatcher.Enabled = true
 	if err := r.reconcileConfigMap(ctx, instance); err != nil {
 		return err
 	}
```
[the remark this answers is not in the document]
Comment (hdkshingala) [reply operator/controllers/controller/reconciler.go:336]: It is intentional as CR watcher need to be enabled when controller is installed using Operator

File: manifests/charts/aperture-controller/README.md [hunk 10]
```diff
@@ -102,64 +102,68 @@
...
-| `controller.config.prometheus.address`                       | specifies the address of the Prometheus server. Example, "http://prometheus-server:80". This must not be empty when prometheus.enabled is set to false. | `nil`    |
+| Name                                                         | Description                                                                                                                                                               | Value                 |
+| ------------------------------------------------------------ | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | --------------------- |
+| `controller.namespaceScoped`                                 | Specifies whether the Controller installation should be namespace-scoped, avoiding the creation of cluster-level resources. This takes precedence over controller.create. | `false`               |
```
[the remark this answers is not in the document]
Comment (hdkshingala) [reply manifests/charts/aperture-controller/README.md:107]: This chart is only for Controller and not for agent

File: cmd/aperturectl/cmd/apply/policy.go [hunk 11]
```diff
@@ -15,7 +15,7 @@ import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/client-go/kubernetes"
 	"k8s.io/client-go/kubernetes/scheme"
-	"sigs.k8s.io/controller-runtime/pkg/client"
+	k8sclient "sigs.k8s.io/controller-runtime/pkg/client"
```
[the remark this answers is not in the document]
Comment (hdkshingala) [reply cmd/aperturectl/cmd/apply/policy.go:18]: client is used as name for other variable so this is intentional and fine

File: docs/content/reference/configuration/controller.md [hunk 12]
```diff
@@ -175,6 +175,14 @@ Env-Var Prefix: `APERTURE_CONTROLLER_OTEL_`
...
 
+<dt>cr_watcher</dt>
+<dd>
+
+([CRWatcherConfig](#c-r-watcher-config))
+Env-Var Prefix: `APERTURE_CONTROLLER_POLICIES_CR_WATCHER_`
+
+</dd>
+
 <dt>promql_jobs_scheduler</dt>
```
[the remark this answers is not in the document]
Comment (hdkshingala) [reply docs/content/reference/configuration/controller.md:186]: This file is auto generated and can't be changed

File: docs/content/reference/configuration/agent.md [hunk 13]
```diff
@@ -526,6 +526,12 @@ Key in this map refers to OTEL pipeline name. Prefixing pipeline name with `metr
 is optional, as all the components and pipeline names would be normalized.
 By default `kubeletstats` custom metrics is added, which can be overwritten.
 
+Below is example to overwrite `kubeletstats` custom metrics:
+
+    otel:
+    	custom_metrics:
+    		kubeletstats: {}
```
[the remark this answers is not in the document]
Comment (hdkshingala) [reply docs/content/reference/configuration/agent.md:533]: It is already properly formatted

File: cmd/aperturectl/cmd/apply/policy.go [hunk 14]
```diff
@@ -133,27 +134,42 @@ func createAndApplyPolicy(policy *policyv1alpha1.Policy) error {
...
+			}
+		} else {
+			return fmt.Errorf("failed to apply policy in Kubernetes: %w", err)
+		}
 	}
 
-	log.Info().Str("policy", policy.GetName()).Str("namespace", policy.GetNamespace()).Msg("Applied Policy successfully")
+	log.Info().Str("policy", name).Str("namespace", deployment.GetNamespace()).Msg("Applied Policy successfully")
 	return nil
 }
```
[the remark this answers is not in the document]
Comment (hdkshingala) [reply cmd/aperturectl/cmd/apply/policy.go:174]: That client is defined as part of the init part of this and is available

File: cmd/aperture-agent/config/types.go [hunk 15]
```diff
@@ -61,6 +61,13 @@ type AgentOTELConfig struct {
...
 	// is optional, as all the components and pipeline names would be normalized.
 	// By default `kubeletstats` custom metrics is added, which can be overwritten.
+	//
+	// Below is example to overwrite `kubeletstats` custom metrics:
+	//
+	//	otel:
+	//		custom_metrics:
+	//			kubeletstats: {}
+	//
 	CustomMetrics map[string]CustomMetricsConfig `json:"custom_metrics,omitempty"`
```
[the remark this answers is not in the document]
Comment (hdkshingala) [reply cmd/aperture-agent/config/types.go:71]: This example is to overwrite to disable the metrics which is correct

File: pkg/policies/controlplane/validator.go [hunk 16]
```diff
@@ -22,13 +23,31 @@ type FxOut struct {
...
+	err := in.Unmarshaller.UnmarshalKey(crwatcher.ConfigKey, &config)
+	if err != nil {
+		log.Error().Err(err).Msg("Failed to unmarshal Kubernetes watcher config")
+		return FxOut{}, err
+	}
+
+	if !config.Enabled {
+		log.Info().Msg("Kubernetes watcher is disabled")
+		return FxOut{}, nil
+	}
```
[the remark this answers is not in the document]
Comment (hdkshingala) [reply pkg/policies/controlplane/validator.go:46]: By default it returns nil and the different log is needed

File: pkg/policies/controlplane/validator.go [hunk 17]
```diff
@@ -10,6 +10,7 @@ import (
 	"github.com/fluxninja/aperture/pkg/config"
 	"github.com/fluxninja/aperture/pkg/log"
 	"github.com/fluxninja/aperture/pkg/policies/controlplane/circuitfactory"
+	"github.com/fluxninja/aperture/pkg/policies/controlplane/crwatcher"
 	"github.com/fluxninja/aperture/pkg/policies/flowcontrol/resources/classifier/compiler"
 	"github.com/fluxninja/aperture/pkg/status"
 	"github.com/fluxninja/aperture/pkg/webhooks/policyvalidator"
```
[the remark this answers is not in the document]
Comment (hdkshingala) [reply pkg/policies/controlplane/validator.go:16]: It is used to get the config

File: playground/Tiltfile [hunk 18]
```diff
@@ -890,7 +880,7 @@ def process_aperture_scenario(scenario_path):
             "child_resources": [
                 {
                     "new_name": "scenario-policies",
-                    "resource_deps": ["controller-service"],
+                    "resource_deps": ["controller-service", "agent"],
                     "extra_objects": [
                         (None, "Policy", APERTURE_CONTROLLER_NS)
                     ]
```
[the remark this answers is not in the document]
Comment (hdkshingala) [reply playground/Tiltfile:886]: This is intentional

File: pkg/policies/controlplane/provider.go [hunk 19]
```diff
@@ -18,14 +30,155 @@ const (
...
+		OnStop: func(ctx context.Context) error {
+			return multierr.Combine(
+				trackers.Stop(),
+				dynamicConfigTrackers.Stop(),
+			)
+		},
+	})
+
+	return trackers, dynamicConfigTrackers, nil
+}
```
[the remark this answers is not in the document]
Comment (hdkshingala) [reply pkg/policies/controlplane/provider.go:89]: return multierr.Combine() returns the error to parent

File: pkg/agentfunctions/config/config.go [hunk 20]
```diff
@@ -5,11 +5,19 @@ import (
 	grpcclient "github.com/fluxninja/aperture/pkg/net/grpc"
 )
 
+// swagger:operation POST /agent_functions common-configuration Agent Functions
+// ---
+// x-fn-config-env: true
+// parameters:
+// - in: body
+//   schema:
+//     $ref: "#/definitions/AgentFunctionsConfig"
```
[the remark this answers is not in the document]
Comment (hdkshingala) [reply pkg/agentfunctions/config/config.go:14]: This is not required and in sync with other places

File: cmd/aperturectl/cmd/apply/root.go [hunk 21]
```diff
@@ -32,6 +36,39 @@ Use this command to apply the Aperture Policies.`,
...
+		err = controller.PreRunE(cmd, args)
+		if err != nil {
+			return err
+		}
+
+		client, err = controller.Client()
+		if err != nil {
+			return err
+		}
 		return nil
```
[the remark this answers is not in the document]
Comment (hdkshingala) [reply cmd/aperturectl/cmd/apply/root.go:71]: The errors received here already have enough context so no need of wrapping them

File: cmd/aperturectl/cmd/apply/root.go [hunk 22]
```diff
@@ -4,16 +4,20 @@ import (
...
 	kubeConfig     string
 	kubeRestConfig *rest.Config
+	controller     utils.ControllerConn
+	client         cmdv1.ControllerClient
+	controllerNs   string
 )
 
 func init() {
-	ApplyCmd.Flags().StringVar(&kubeConfig, "kube-config", "", "Path to the Kubernetes cluster config. Defaults to '~/.kube/config' or $KUBECONFIG")
+	controller.InitFlags(ApplyCmd.PersistentFlags())
```
[the remark this answers is not in the document]
Comment (hdkshingala) [reply cmd/aperturectl/cmd/apply/root.go:21]: This is already there

File: cmd/aperturectl/cmd/blueprints/generate.go [hunk 23]
```diff
@@ -152,6 +154,11 @@ aperturectl blueprints generate --name=policies/static-rate-limiting --values-fi
 				return err
 			}
 
+			err = apply.ApplyPolicyCmd.Flag("kube").Value.Set("true")
```
[the remark this answers is not in the document]
Comment (hdkshingala) [reply cmd/aperturectl/cmd/blueprints/generate.go:157]: This is correct as we want to enable kube flag here

File: api/aperture/policy/language/v1/policy.proto [hunk 24]
```diff
@@ -10,15 +10,80 @@ import "aperture/policy/language/v1/std_components.proto";
 import "google/api/annotations.proto";
 import "google/protobuf/duration.proto";
 import "google/protobuf/empty.proto";
+import "google/protobuf/struct.proto";
 import "protoc-gen-openapiv2/options/annotations.proto";
```
[the remark this answers is not in the document]
Comment (hdkshingala) [reply api/aperture/policy/language/v1/policy.proto:14]: Indentation is already correct

File: cmd/aperture-agent/config/types.go [hunk 25]
```diff
@@ -52,6 +52,8 @@ import (
 //swagger:model
 type AgentOTELConfig struct {
 	otelconfig.CommonOTELConfig `json:",inline"`
+	// DisableKubernetesScraper disables metrics collection for Kubernetes resources.
+	DisableKubernetesScraper bool `json:"disable_kubernetes_scraper" default:"false"`
```
[the remark this answers is not in the document]
Comment (hdkshingala) [reply cmd/aperture-agent/config/types.go:56]: The suggested changes are same as existing code

File: manifests/charts/aperture-controller/README.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (hdkshingala) [reply manifests/charts/aperture-controller/README.md:107]: This is incorrect suggestion. create and namespacedScoped are 2 different flags and namespaceScoped takes precedence over create

**Review by kwapik — APPROVED** [review]

File: operator/controllers/controller/configmaps.go [hunk 26]
```diff
@@ -68,6 +69,27 @@ func configMapForControllerConfig(instance *controllerv1alpha1.Controller, schem
...
+	}
+
+	if err := ctrl.SetControllerReference(instance, cm, scheme); err != nil {
+		return nil, err
+	}
+
+	return cm, nil
+}
+
 // createConfigMap calls the Kubernetes API to create the provided Controller ConfigMap resource.
```
[the remark this answers is not in the document]
Comment (hdkshingala) [reply operator/controllers/controller/configmaps.go:93]: This is a client certificate and ok to store in configmap

File: pkg/policies/controlplane/policy-factory.go [hunk 27]
```diff
@@ -12,50 +12,68 @@ import (
...
+var (
+	policiesEtcdWatcherFxTag              = "policies-driver"
+	policiesDynamicConfigEtcdWarcherFxTag = "policies-dynamic-config-driver"
+)
+
 // policyFactoryModule module for policy factory.
 func policyFactoryModule() fx.Option {
 	return fx.Options(
+		etcdwatcher.Constructor{Name: policiesEtcdWatcherFxTag, EtcdPath: paths.PoliciesConfigPath}.Annotate(),
+		etcdwatcher.Constructor{Name: policiesDynamicConfigEtcdWarcherFxTag, EtcdPath: paths.PoliciesDynamicConfigPath}.Annotate(),
```
[the remark this answers is not in the document]
Comment (hdkshingala) [reply pkg/policies/controlplane/policy-factory.go:35]: The suggested and actual names are same i.e no typos

**Pull request closed without merging**